src/common/collections.rs

```diff
@@ -41,6 +43,42 @@ pub async fn do_list_collections(toc: &TableOfContent) -> CollectionsResponse {
     CollectionsResponse { collections }
 }
 
+/// Construct shards-replicas layout for the shard from the given scope of peers
+/// Example:
+///   Shards: 3
+///   Replicas: 2
+///   Peers: [A, B, C]
+///
+/// Placement:
+/// [
+///         [A, B]
+///         [B, C]
+///         [A, C]
+/// ]
+fn generate_even_placement(
+    mut pool: Vec<PeerId>,
+    shard_number: usize,
+    replication_factor: usize,
+) -> Vec<Vec<PeerId>> {
+    let mut exact_placement = Vec::new();
+    let mut rng = rand::thread_rng();
+    pool.shuffle(&mut rng);
+    let mut next_pool = pool.clone();
+
+    let max_replication_factor = std::cmp::min(replication_factor, pool.len());
```

Comment:
replace with looped iterator

Reply:
Nice!

What about full iterator style? This also removes the `Vec::new()`'s and allocates the vectors with the right capacity right away.

```rust
fn generate_even_placement(
    mut pool: Vec<PeerId>,
    shard_number: usize,
    replication_factor: usize,
) -> ShardsPlacement {
    pool.shuffle(&mut rand::thread_rng());
    let max_replication_factor = std::cmp::min(replication_factor, pool.len());
    let mut pool = pool.iter().cycle();

    // pool: [1,2,3,4]
    // shuf_pool: [2,3,4,1]
    //
    // loop_iter:       [2, 3, 4, 1, 2, 3, 4, 1, 2, 3, 4, 1, 2, 3, 4, 1,...]
    // shard_placement: [2, 3, 4][1, 2, 3][4, 1, 2][3, 4, 1][2, 3, 4]

    std::iter::repeat_with(|| {
        pool.by_ref()
            .take(max_replication_factor)
            .copied()
            .collect()
    })
    .take(shard_number)
    .collect()
}
```